File: numberstepper/src/main/java/com/maple/numberstepper/NumberStepper.java
package com.maple.numberstepper;

import android.animation.Animator;
import android.animation.ValueAnimator;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewConfiguration;
import android.view.animation.AccelerateInterpolator;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import java.lang.ref.WeakReference;


/**
 * 数字 加减 步进器
 *
 * @author : shaoshuai
 * @date ：2021/2/3
 */
public class NumberStepper extends RelativeLayout implements View.OnTouchListener {
    private NumberStepperValueChangeListener listener;
    private TextView tvStepperContent;
    private ImageView ivStepperMinus, ivStepperPlus;
    public static int ANIMATION_DURATION = 300;//恢复动画时间
    public boolean animationing = false;//动画中,不能进行滑动
    private UpdateRunnable updateRunnable;
    private boolean stepTouch = false;//是否按着，判断是否还要继续更新数值和界面

    //按下后多少间隔触发快速改变模式
    private static final long STEP_SPEED_CHANGE_DURATION = 1000;
    private static final long UPDATE_DURATION_SLOW = 300;//数值更新频率-慢
    private static final long UPDATE_DURATION_FAST = 100;//数值更新频率-快
    private int valueSlowStep = 1;//慢速递增值 步长

    //按下的初始x值
    private float startX = 0;
    private float startStepperContentLeft = 0;
    private boolean hasStepperContentLeft = false;
    //按下时间
    private long startTime = 0;

    //当前状态
    private int status = STATUS_NORMAL;
    private static final int STATUS_MIMNUS = -1;
    private static final int STATUS_PLUS = 1;
    private static final int STATUS_NORMAL = 0;
    //当前模式
    private Mode mode = Mode.AUTO;

    public enum Mode {
        AUTO(0), CUSTOM(1);
        private final int value;

        Mode(int value) {
            this.value = value;
        }

        public int getValue() {
            return value;
        }

        public static Mode valueOf(int value) {    // 手写的从int到enum的转换函数
            switch (value) {
                case 0:
                    return AUTO;
                case 1:
                    return CUSTOM;
            }
            return AUTO;
        }
    }

    private int value = 0;
    private int minValue = 0;
    private int maxValue = 100;

    public NumberStepper(Context context) {
        this(context, null);
    }

    public NumberStepper(Context context, AttributeSet attrs) {
        super(context, attrs);
        initViews(attrs);
    }

    private void initViews(AttributeSet attrs) {
        LayoutInflater.from(getContext()).inflate(R.layout.ms_view_number_stepper, this, true);
        tvStepperContent = (TextView) findViewById(R.id.tvStepperContent);
        ivStepperMinus = (ImageView) findViewById(R.id.ivStepperMinus);
        ivStepperPlus = (ImageView) findViewById(R.id.ivStepperPlus);

        String text = "";
        Drawable background = null;
        Drawable contentBackground = null;
        Drawable leftButtonResources = null;
        Drawable rightButtonResources = null;
        Drawable leftButtonBackground = null;
        Drawable rightButtonBackground = null;
        int contentTextColor = getResources().getColor(R.color.ms_stepper_text);
        float contentTextSize = 0;
        if (attrs != null) {
            TypedArray a = getContext().obtainStyledAttributes(attrs, R.styleable.NumberStepper);
            int modeValue = a.getInt(R.styleable.NumberStepper_mode, Mode.AUTO.getValue());
            mode = Mode.valueOf(modeValue);
            minValue = a.getInt(R.styleable.NumberStepper_min, minValue);
            maxValue = a.getInt(R.styleable.NumberStepper_max, maxValue);
            value = valueRangeCheck(a.getInt(R.styleable.NumberStepper_value, value));
            valueSlowStep = a.getInt(R.styleable.NumberStepper_step, valueSlowStep);
            if (valueSlowStep <= 0) valueSlowStep = 1;
            text = a.getString(R.styleable.NumberStepper_text);

            background = a.getDrawable(R.styleable.NumberStepper_stepper_background);
            contentBackground = a.getDrawable(R.styleable.NumberStepper_stepper_contentBackground);
            leftButtonResources = a.getDrawable(R.styleable.NumberStepper_stepper_leftButtonResources);
            rightButtonResources = a.getDrawable(R.styleable.NumberStepper_stepper_rightButtonResources);
            leftButtonBackground = a.getDrawable(R.styleable.NumberStepper_stepper_leftButtonBackground);
            rightButtonBackground = a.getDrawable(R.styleable.NumberStepper_stepper_rightButtonBackground);

            contentTextColor = a.getColor(R.styleable.NumberStepper_stepper_contentTextColor, contentTextColor);

            contentTextSize = a.getFloat(R.styleable.NumberStepper_stepper_contentTextSize, 0);
            a.recycle();
        }

        if (background != null) {
            setBackground(background);
        } else {
            setBackgroundResource(R.color.ms_stepper_button_press);
        }


        if (contentBackground != null) {
            setContentBackground(contentBackground);
        }
        tvStepperContent.setTextColor(contentTextColor);
        if (contentTextSize > 0)
            setContentTextSize(contentTextSize);

        if (leftButtonBackground != null) {
            ivStepperMinus.setBackground(leftButtonBackground);
        }
        if (rightButtonBackground != null) {
            ivStepperPlus.setBackground(rightButtonBackground);
        }

        if (leftButtonResources != null) {
            setLeftButtonResources(leftButtonResources);
        }
        if (rightButtonResources != null) {
            setRightButtonResources(rightButtonResources);
        }

        if (mode == Mode.AUTO)//AUTO模式，写数值到滑动条上
            tvStepperContent.setText(String.valueOf(value));
        else
            tvStepperContent.setText(text);

        //设置后onclick产生的点击状态会失效
        ivStepperMinus.setOnTouchListener(this);
        ivStepperPlus.setOnTouchListener(this);
        setOnTouchListener(this);

        updateRunnable = new UpdateRunnable(this);
    }

    @Override
    public boolean onTouch(View v, MotionEvent event) {
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                stepTouch = true;
                postDelayed(updateRunnable, UPDATE_DURATION_SLOW);
                //非按钮则记录位置
                startX = event.getX();
                initStartStepperContentLeft();
                startTime = System.currentTimeMillis();
                //如果是两边的按钮，分别设置为点击状态
                if (v == ivStepperMinus) {
                    ivStepperMinus.setPressed(true);
                    status = STATUS_MIMNUS;
                    break;
                } else if (v == ivStepperPlus) {
                    ivStepperPlus.setPressed(true);
                    status = STATUS_PLUS;
                    break;
                }
                break;
            case MotionEvent.ACTION_MOVE:
                //是按钮则不能移动,恢复位置的动画中也不能移动
                if (v == ivStepperMinus || v == ivStepperPlus || animationing) break;
                //非按钮则进行移动
                float moveX = event.getX() - startX;
                float x = moveX + startStepperContentLeft;
                moveStepperContent(x);
                moveEffectStatus(moveX);
                break;
            case MotionEvent.ACTION_UP:
            case MotionEvent.ACTION_CANCEL:
                stepTouch = false;
                //如果是两边的按钮，分别设置为点击状态
                if (v == ivStepperMinus) {
                    ivStepperMinus.setPressed(false);
                    break;
                } else if (v == ivStepperPlus) {
                    ivStepperPlus.setPressed(false);
                    break;
                }
                restoreStepperContent();
                break;
        }
        return true;
    }

    private void initStartStepperContentLeft() {
        if (hasStepperContentLeft) return;
        hasStepperContentLeft = true;
        startStepperContentLeft = tvStepperContent.getLeft();
    }

    /**
     * 中间滑条恢复原位置
     */
    private void restoreStepperContent() {
        if (animationing) return;
        animationing = true;
        ValueAnimator restoreTranslateAnimation = ValueAnimator.ofFloat(tvStepperContent.getLeft(), (int) startStepperContentLeft);
        restoreTranslateAnimation.setDuration(ANIMATION_DURATION);
        restoreTranslateAnimation.addListener(new Animator.AnimatorListener() {

            @Override
            public void onAnimationStart(Animator animation) {
            }

            @Override
            public void onAnimationEnd(Animator animation) {
                animationing = false;
            }

            @Override
            public void onAnimationCancel(Animator animation) {
            }

            @Override
            public void onAnimationRepeat(Animator animation) {
            }
        });
        restoreTranslateAnimation.addUpdateListener(animation -> {
            Float value = (Float) animation.getAnimatedValue();
            moveStepperContent(value);
        });
        restoreTranslateAnimation.setInterpolator(new AccelerateInterpolator());
        restoreTranslateAnimation.start();
    }

    /**
     * 移动位置
     */
    private void moveStepperContent(float x) {
        LayoutParams params = new LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT);
        params.leftMargin = (int) x;
        //限制子控件移动必须在视图范围内
        if (params.leftMargin < 0 || (params.leftMargin + tvStepperContent.getWidth()) > getWidth())
            return;
        params.topMargin = 0;
        params.width = tvStepperContent.getWidth();
        params.height = tvStepperContent.getHeight();

        tvStepperContent.setLayoutParams(params);
    }


    private void moveEffectStatus(float x) {
        int scaledTouchSlop = ViewConfiguration.get(getContext()).getScaledTouchSlop();
        if (x > scaledTouchSlop) {
            status = STATUS_PLUS;
        } else if (x < -scaledTouchSlop) {
            status = STATUS_MIMNUS;
        } else {
            status = STATUS_NORMAL;
        }
    }

    private int getNextValue() {
        switch (status) {
            case STATUS_MIMNUS:
                return value - valueSlowStep;
            case STATUS_PLUS:
                return value + valueSlowStep;
            case STATUS_NORMAL:
                return value;
        }
        return value;
    }

    private void updateUI() {
        int nextValue = getNextValue();
        //判断是否在范围之内
        if (nextValue < minValue) {
            nextValue = minValue;
        }
        if (nextValue > maxValue) {
            nextValue = maxValue;
        }
        value = nextValue;
        if (mode == Mode.AUTO)//AUTO模式，写数值到滑动条上
            tvStepperContent.setText(String.valueOf(value));
        if (listener != null)
            listener.onValueChange(this, value);
        if (stepTouch)
            postDelayed(updateRunnable, (System.currentTimeMillis() - startTime > STEP_SPEED_CHANGE_DURATION) ? UPDATE_DURATION_FAST : UPDATE_DURATION_SLOW);
    }


    static class UpdateRunnable implements Runnable {
        private WeakReference<NumberStepper> view;

        public UpdateRunnable(NumberStepper view) {
            this.view = new WeakReference<NumberStepper>(view);
        }

        public void run() {
            NumberStepper stepper = view.get();
            if (stepper != null) {
                stepper.updateUI();
            }
        }
    }

    public interface NumberStepperValueChangeListener {
        public void onValueChange(View view, int value);
    }

    public void setOnValueChangeListener(NumberStepperValueChangeListener listener) {
        this.listener = listener;
    }

    /**
     * 返回当前模式类型
     *
     * @return
     */
    public Mode getMode() {
        return mode;
    }

    /**
     * 模式设置 AUTO(0) 数值写到滑动条, CUSTOM(1) 自定义文字;
     *
     * @param mode
     */
    public void setMode(Mode mode) {
        this.mode = mode;
    }

    /**
     * 获取当前值
     *
     * @return
     */
    public int getValue() {
        return value;
    }

    /**
     * 设置当前值
     *
     * @param value
     */
    public void setValue(int value) {
        this.value = valueRangeCheck(value);
        if (mode == Mode.AUTO)//AUTO模式，写数值到滑动条上
            tvStepperContent.setText(String.valueOf(value));
    }

    public int valueRangeCheck(int value) {
        if (value > maxValue) value = maxValue;
        else if (value < minValue) value = minValue;
        return value;
    }


    public int getMinValue() {
        return minValue;
    }


    public void setMinValue(int minValue) {
        this.minValue = minValue;
    }

    public int getMaxValue() {
        return maxValue;
    }

    public void setMaxValue(int maxValue) {
        this.maxValue = maxValue;
    }

    /**
     * 获取步长
     */
    public int getValueSlowStep() {
        return valueSlowStep;
    }

    /**
     * 设置步长
     */
    public void setValueSlowStep(int valueSlowStep) {
        this.valueSlowStep = valueSlowStep;
    }

    /**
     * 设置中间内容滑条颜色
     */
    public void setContentBackground(int resId) {
        tvStepperContent.setBackgroundResource(resId);
    }

    public void setContentBackground(Drawable drawable) {
        tvStepperContent.setBackground(drawable);
    }

    /**
     * 设置中间内容文字颜色
     */
    public void setContentTextColor(int resId) {
        tvStepperContent.setTextColor(getResources().getColor(resId));
    }

    /**
     * 设置中间内容文字,mode需为Custom才支持
     */
    public void setText(String text) {
        tvStepperContent.setText(text);
    }

    /**
     * 设置中间内容文字大小
     */
    public void setContentTextSize(float size) {
        tvStepperContent.setTextSize(size);
    }

    /**
     * 设置按钮背景
     */
    public void setButtonBackGround(int resId) {
        ivStepperMinus.setBackgroundResource(resId);
        ivStepperPlus.setBackgroundResource(resId);
    }

    /**
     * 设置按钮资源
     */
    public void setLeftButtonResources(int resId) {
        ivStepperMinus.setImageResource(resId);
    }

    /**
     * 设置按钮资源
     */
    public void setLeftButtonResources(Drawable drawable) {
        ivStepperMinus.setImageDrawable(drawable);
    }

    /**
     * 设置按钮资源
     */
    public void setRightButtonResources(int resId) {
        ivStepperPlus.setImageResource(resId);
    }

    /**
     * 设置按钮资源
     */
    public void setRightButtonResources(Drawable drawable) {
        ivStepperPlus.setImageDrawable(drawable);
    }
}

File: numberstepper/src/main/java/com/maple/numberstepper/NumberStepper.kt
package com.maple.numberstepper

import android.animation.Animator
import android.animation.ValueAnimator
import android.content.Context
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.View.OnTouchListener
import android.view.ViewConfiguration
import android.view.animation.AccelerateInterpolator
import android.widget.ImageView
import android.widget.RelativeLayout
import android.widget.TextView
import androidx.core.content.ContextCompat
import com.maple.numberstepper.DensityUtils.dp2px
import java.lang.ref.WeakReference
import kotlin.math.max
import kotlin.math.min

/**
 * 数字 加减 步进器
 *
 * @author : shaoshuai
 * @date ：2021/2/3
 */
class NumberStepper : RelativeLayout, OnTouchListener {
    lateinit var tvText: TextView
    lateinit var ivLeft: ImageView
    lateinit var ivRight: ImageView

    var contentFinalText: String? = null // 固定文本
    var minValue = 0
    var maxValue = 100
    var stepValue = 1 // 步长：每次增减数
    var currentValue = 0
        set(value) {
            field = max(minValue, min(maxValue, value))
            if (contentFinalText.isNullOrEmpty()) {
                tvText.text = value.toString()
            }
        }

    companion object {
        private const val STEP_SPEED_CHANGE_DURATION: Long = 1000 //按下后多少间隔触发快速改变模式
        private const val UPDATE_DURATION_SLOW: Long = 300 //数值更新频率-慢
        private const val UPDATE_DURATION_FAST: Long = 100 //数值更新频率-快
        private const val STATUS_MIMNUS = -1
        private const val STATUS_PLUS = 1
        private const val STATUS_NORMAL = 0
    }

    constructor(context: Context) : this(context, null)
    constructor(context: Context, attrs: AttributeSet?) : this(context, attrs, 0)
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        initViews(context, attrs)
    }

    private fun initViews(context: Context, attrs: AttributeSet?) {
        val view = LayoutInflater.from(context).inflate(R.layout.ms_view_number_stepper, this, true)
        tvText = view.findViewById(R.id.tv_text)
        ivLeft = view.findViewById(R.id.iv_left)
        ivRight = view.findViewById(R.id.iv_right)
        var viewBackground: Drawable? = null
        var contentBackground: Drawable? = null
        var contentTextSize = 15f
        var contentTextColor = ContextCompat.getColor(context, R.color.ms_stepper_text)
        var buttonWidth = 30f.dp2px(context)
        var leftButtonBackground: Drawable? = null
        var leftButtonResources: Drawable? = null
        var rightButtonBackground: Drawable? = null
        var rightButtonResources: Drawable? = null
        if (attrs != null) {
            val a = context.obtainStyledAttributes(attrs, R.styleable.NumberStepper)
            maxValue = a.getInt(R.styleable.NumberStepper_maxValue, maxValue)
            minValue = a.getInt(R.styleable.NumberStepper_minValue, minValue)
            currentValue = a.getInt(R.styleable.NumberStepper_currentValue, currentValue)
            stepValue = a.getInt(R.styleable.NumberStepper_stepValue, stepValue)
            viewBackground = a.getDrawable(R.styleable.NumberStepper_stepper_background)
            contentBackground = a.getDrawable(R.styleable.NumberStepper_stepper_contentBackground)
            contentFinalText = a.getString(R.styleable.NumberStepper_stepper_final_text)
            contentTextColor = a.getColor(R.styleable.NumberStepper_stepper_contentTextColor, contentTextColor)
            contentTextSize = a.getFloat(R.styleable.NumberStepper_stepper_contentTextSize, contentTextSize)
            buttonWidth = a.getDimensionPixelSize(R.styleable.NumberStepper_stepper_buttonWidth, buttonWidth)
            leftButtonBackground = a.getDrawable(R.styleable.NumberStepper_stepper_leftButtonBackground)
            leftButtonResources = a.getDrawable(R.styleable.NumberStepper_stepper_leftButtonResources)
            rightButtonBackground = a.getDrawable(R.styleable.NumberStepper_stepper_rightButtonBackground)
            rightButtonResources = a.getDrawable(R.styleable.NumberStepper_stepper_rightButtonResources)
            a.recycle()
        }
        if (viewBackground != null) {
            background = viewBackground
        } else {
            setBackgroundResource(R.color.ms_stepper_button_press)
        }
        with(tvText) {
            contentBackground?.let { background = it }
            setTextColor(contentTextColor)
            textSize = contentTextSize
            contentFinalText?.let { text = it }
        }
        with(ivLeft) {
            leftButtonBackground?.let { background = it }
            leftButtonResources?.let { setImageDrawable(it) }
            layoutParams = layoutParams.apply { width = buttonWidth }
        }
        with(ivRight) {
            rightButtonBackground?.let { background = it }
            rightButtonResources?.let { setImageDrawable(it) }
            layoutParams = layoutParams.apply { width = buttonWidth }
        }
        //设置后onclick产生的点击状态会失效
        ivLeft.setOnTouchListener(this)
        ivRight.setOnTouchListener(this)
        setOnTouchListener(this)
    }


    private var startX = 0f//按下的初始x值
    private var startTime: Long = 0//按下时间
    private var status = STATUS_NORMAL//当前状态
    private var stepTouch = false //是否按着，判断是否还要继续更新数值和界面
    private var startStepperContentLeft = 0f
    private var hasStepperContentLeft = false

    override fun onTouch(v: View, event: MotionEvent): Boolean {
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                stepTouch = true
                postDelayed(updateRunnable, UPDATE_DURATION_SLOW)
                //非按钮则记录位置
                startX = event.x
                startTime = System.currentTimeMillis()
                if (!hasStepperContentLeft) {
                    hasStepperContentLeft = true
                    startStepperContentLeft = tvText.left.toFloat()
                }
                //如果是两边的按钮，分别设置为点击状态
                if (v === ivLeft) {
                    ivLeft.isPressed = true
                    status = STATUS_MIMNUS
                } else if (v === ivRight) {
                    ivRight.isPressed = true
                    status = STATUS_PLUS
                }
            }
            MotionEvent.ACTION_MOVE -> {
                //是按钮则不能移动,恢复位置的动画中也不能移动
                if (v == ivLeft || v == ivRight || animationing) {
                    // break
                } else {
                    //非按钮则进行移动
                    val moveX = event.x - startX
                    val x = moveX + startStepperContentLeft
                    moveStepperContent(x)
                    // moveEffectStatus(moveX)
                    val scaledTouchSlop = ViewConfiguration.get(context).scaledTouchSlop
                    status = if (moveX > scaledTouchSlop) {
                        STATUS_PLUS
                    } else if (moveX < -scaledTouchSlop) {
                        STATUS_MIMNUS
                    } else {
                        STATUS_NORMAL
                    }
                }
            }
            MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
                stepTouch = false
                //如果是两边的按钮，分别设置为点击状态
                when (v) {
                    ivLeft -> ivLeft.isPressed = false
                    ivRight -> ivRight.isPressed = false
                    else -> restoreStepperContent()
                }
            }
        }
        return true
    }


    private var animationing = false //动画中,不能进行滑动
    private val updateRunnable: UpdateRunnable by lazy { UpdateRunnable(this) }

    /**
     * 中间滑条恢复原位置
     */
    private fun restoreStepperContent() {
        if (animationing)
            return
        animationing = true
        ValueAnimator.ofFloat(tvText.left.toFloat(), startStepperContentLeft).apply {
            duration = 300L // 恢复动画时间
            interpolator = AccelerateInterpolator()
            addListener(object : Animator.AnimatorListener {
                override fun onAnimationCancel(animation: Animator) {}
                override fun onAnimationRepeat(animation: Animator) {}
                override fun onAnimationStart(animation: Animator) {}
                override fun onAnimationEnd(animation: Animator) {
                    animationing = false
                }
            })
            addUpdateListener { animation: ValueAnimator ->
                val value = animation.animatedValue as Float
                moveStepperContent(value)
            }
        }.start()
    }

    /**
     * 移动位置
     */
    private fun moveStepperContent(x: Float) {
        //限制子控件移动必须在视图范围内
        if (x < 0 || x + tvText.width > width)
            return
        val params = LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT)
        params.leftMargin = x.toInt()
        params.topMargin = 0
        params.width = tvText.width
        params.height = tvText.height
        tvText.layoutParams = params
    }

    internal class UpdateRunnable(view: NumberStepper) : Runnable {
        private val view: WeakReference<NumberStepper> = WeakReference(view)
        override fun run() {
            val stepper = view.get()
            stepper?.updateUI()
        }
    }

    private fun updateUI() {
        currentValue = getNextValue()
        if (contentFinalText.isNullOrEmpty()) {
            tvText.text = currentValue.toString()
        }
        listener?.onValueChange(this, currentValue)
        if (stepTouch) {
            val duration = if (System.currentTimeMillis() - startTime > STEP_SPEED_CHANGE_DURATION) UPDATE_DURATION_FAST else UPDATE_DURATION_SLOW
            postDelayed(updateRunnable, duration)
        }
    }

    private fun getNextValue(): Int = when (status) {
        STATUS_MIMNUS -> currentValue - stepValue
        STATUS_PLUS -> currentValue + stepValue
        // STATUS_NORMAL
        else -> currentValue
    }

    private var listener: NumberStepperValueChangeListener? = null

    interface NumberStepperValueChangeListener {
        fun onValueChange(view: View, value: Int)
    }

    fun setOnValueChangeListener(listener: NumberStepperValueChangeListener?) {
        this.listener = listener
    }
}